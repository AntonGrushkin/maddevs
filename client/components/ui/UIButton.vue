<template>
  <button
    class="ui-button"
    :class="{ 'ui-button--disabled': disabled }"
    @click="onClick"
    v-WaveAnimation="!disabled"
  >{{ name }}</button>
</template>

<script>
import WaveAnimation from '@/directives/WaveAnimation';

export default {
  name: 'UIButton',
  directives: {
    WaveAnimation
  },
  props: {
    name: {
      type: String,
      default: 'Button'
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  methods: {
    onClick() {
      if (this.$props.disabled === false) {
        this.$emit('click');
      }
    }
  }
};
</script>

<style lang="scss" scoped>
  @import '../../assets/styles/vars';

  .ui-button {
    height: 64px;
    border: 1px solid $border-color--red;
    background-color: $bgcolor--red;
    color: $text-color--white;
    font-size: 21px;
    font-family: 'Hoves-Bold';
    letter-spacing: -0.03em;
    box-shadow: none;
    transition: 0.2s;
    cursor: pointer;

    &:not(.ui-button--disabled) {
      &:hover {
        background-color: $bgcolor--red;
        color: $text-color--white;
      }

      &:active {
        background-color: $button-active--red;
        border-color: $button-active-border--red;
      }
    }

    &--disabled {
      cursor: not-allowed;
      opacity: 0.5;
      user-select: none;
    }

    &--transparent-bgc {
      background-color: transparent;
      color: $text-color--red;
    }
  }
</style>
