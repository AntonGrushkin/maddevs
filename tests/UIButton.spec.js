import {
  mount
} from '@vue/test-utils';
import UIButton from '@/components/ui/UIButton';

describe('Ui button modal trigger', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(UIButton, {
      propsData: {
        name: 'Button name',
        disabled: false
      }
    });
  });

  // ------ IMPORTANT ----- //
  test('is a Vue instance', () => {
    expect(wrapper.exists()).toBeTruthy();
  });
  
  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
  // --------------------- //

  test('Button slot renders a default button text', () => {
    const button = wrapper.find('.ui-button');
    expect(button.text()).toBe('Button name');
  });

  test('onClick should call $emit', () => {
    wrapper.vm.onClick();
    expect(wrapper.emitted().click).toBeTruthy();
  });

  test('if this.$props.disabled equal false, onClick should not call $emit and button should contain new class', () => {
    wrapper = mount(UIButton, {
      propsData: {
        name: 'Button name',
        disabled: true
      }
    });
    wrapper.vm.onClick();
    expect(wrapper.emitted().click).toBeFalsy();

    const button = wrapper.find('.ui-button');
    expect(button.classes()).toContain('ui-button--disabled');
  });
});
