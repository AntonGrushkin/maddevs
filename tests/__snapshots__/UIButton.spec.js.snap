// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ui button renders correctly 1`] = `
<button
  class="ui-button"
  style="position: relative; overflow: hidden;"
>
  button text
  <div
    class="wave"
  />
</button>
`;
